Kitchen Machine Refactor and Max Item fix

FalseIncarnate opened the pull request:
Refactors the code for Microwaves, Ovens, Grills, and Candy Makers to utilize a shared parent which holds the common procs.
- This means any new additions or fixes will only need to be implemented in one place, rather than 4, to affect them all.
- Also reduces file sizes on the individual machines' files by removing a lot of the duplicate code

Updated the code for the kitchen machines to be more modular and fixed some issues.
- Replaced color macros with span classes
- Replaced hardcoded names with src references
- Each machine has a cook_verbs list which is used to create the in-use message (Microwaving in progress)
- Each machine has variables to define the icon states for the different states of the machine (on, off, broken, dirty, open).
- Changed a few formerly global variables to not be global, as this would cause issues with the shared code (microwave with candy maker recipes for example)

Fixed a bug where the microwave would be considered "full" before having all of the ingredients for a recipe (example: Super bigbite burger)
- Created a new proc to count the number of items and fruits in a recipe to determine the max_n_items value

Map Edits to Cyberiad and associated Z-levels for pathing changes (map-merged)
- MetaStation will need to be updated as well, I'll do this in another PR unless someone else beats me to it.

Krausus commented on _maps/map_files/cyberiad/z2.dmm:
```diff
@@ -748,7 +748,7 @@
 "ou" = (/turf/unsimulated/floor{icon = 'icons/turf/snow.dmi'; icon_state = "snow"},/turf/simulated/shuttle/wall{dir = 1; icon_state = "diagonalWall3"},/area/syndicate_station/start)
 "ov" = (/turf/simulated/shuttle/wall{dir = 4; icon_state = "wall3"},/area/syndicate_station/start)
-"ow" = (/obj/structure/table,/obj/machinery/microwave,/turf/simulated/shuttle/floor{icon_state = "floor4"},/area/syndicate_station/start)
+"ow" = (/obj/structure/table,/obj/item/weapon/circuitboard/microwave,/turf/simulated/shuttle/floor{icon_state = "floor4"},/area/syndicate_station/start)
```
It seems that the microwave on the nuke-ops shuttle turned into a circuit board...

I swear, if any of these microwaves turned into _energy guns..._

Krausus commented on code/game/machinery/kitchen/microwave.dm:
```diff
@@ -46,7 +35,7 @@
 				acceptable_reagents |= reagent
 			if (recipe.items)
-				max_n_of_items = max(max_n_of_items,recipe.items.len)
+				max_n_of_items = max(max_n_of_items,recipe.count_n_items())
```
If the only difference between the four copies of the recipe-loading code is the base type of the recipes, they should be moved to New() on the parent type, with a var for the recipe type to load.

Fox-McCloud commented:
Just need to correct the mapping error and I'd consider this good to go. Ideally, if you implement what Krausus suggests, that would be even better, but I'm not going to make a fuss over it; this is already infinitely better than it used to be.

FalseIncarnate commented:
I thought I got them all, but that one slipped past me. I assure you there are no new energy guns. Though I can definitely make that happen!

I spent most of today asleep or at work, I'll fix the map and look into implementing the suggestion when I get home in an hour or so, shouldn't need to delay past the 24hr mark for those changes

TheDZD commented:
Would it be possible to hold off on this for a bit? Both this and #1476 touch z5 (#1476 made a lot of changes there), so it's likely there's going to be a merge conflicts, and I would prefer not to have to re-do an hour and some's worth of mapping.

Fox-McCloud commented:
Ahhh, good point. Yeah, I'll place a hold on this.

FalseIncarnate commented:
When I'm home, I'll discuss a potential compromise that could allow this to be merged without affecting Z5

It won't be an ideal solution, but should work without much issue

FalseIncarnate commented:
New commit reverts the changes to z5, while still keeping the refactor in place. Microwaves using the old path will automatically replace themselves with the new version when initialized.

Fixed the microwave on the syndi-shuttle de-evolving into a circuit board.

Also implemented some suggestions from Krausus on moving more code into the base level kitchen_machine and out of the individual types' New() procs while cleaning up the code to use fewer loops.

Iamgoofball commented:
Can I ask why you guys just made the candy makers and machines into slightly modified microwaves? It kinda killed the point of them.

FalseIncarnate commented:
The custom food was pretty bad, and very problematic in the case of players being able to make anything into food in a non-reversible manner. Making the candy maker into a modified microwave allowed us to utilize recipes and upgrading for it, standardizing the kitchen equipment and letting us create more options for the crew to cook.

I personally did this in #535 and will gladly answer any questions you have about it, though this PR's comment section may not be the best way of doing so.
